### PR TITLE
Set empty program name to hash of instructions

### LINKED ISSUE #3443: ebpf_program_load_bytes() always calls srand before generating random program name
### Describe the bug

Programs loaded from bytes get a random name assigned if `program_name` is passed as null, but `srand(time(nullptr))` is called before `rand()` in each call, so two programs loaded in the same second will get the same name.

### OS information

_No response_

### Steps taken to reproduce bug

1. Load two ebpf programs from bytes in the same clock second with `program_name` null

### Expected behavior

The generated `unique_name` should be unique, or at least more random.

### Actual outcome

The generated name is the same because rand gets the same seed.

### Additional details

_No response_

### PR DESCRIPTION
**Pull request by mikeagun**
## Description

Resolves #3443 by using the SHA256 hash of the program instructions as the program name instead of using `rand()` to generate the name.

Instruction lists are now given a unique name (fixing the issue), and this solution avoids any potential problems from changing the `rand()` state for the process. Only the first 63 characters of the hash are used to keep under bpf object length limit (64).

As part of this change, hash.cpp and hash.h were moved from tests/libs/util to libs/shared so they are in a more central location for use in other project files.

## Testing

Adds ebpf_program_load_bytes-name-gen test case in end_to_end.cpp

## Documentation

N/A

## Installation

N/A

### CODE REVIEW

**Comment by mikeagun** [comment]
Sorting out signed/unsigned mismatch this caused to make build fail.

File: tests/end_to_end/end_to_end.cpp [hunk 1]
```diff
@@ -2412,6 +2412,62 @@ _create_service_helper(
...
+
+    // Now query the program info and verify it matches what we set.
+    bpf_prog_info program_info = {};
+    uint32_t program_info_size = sizeof(program_info);
+    REQUIRE(bpf_obj_get_info_by_fd(program_fd, &program_info, &program_info_size) == 0);
+    REQUIRE(program_info_size == sizeof(program_info));
+    REQUIRE(program_info.nr_map_ids == 0);
+    REQUIRE(program_info.map_ids == 0);
+    REQUIRE(program_info.name != NULL);
+    // Name should contain SHA256 hash in hex (minus last char to stay under BPF_OBJ_NAME_LEN)
```
Comment (dthaler) [inline tests/end_to_end/end_to_end.cpp:2462]:
```suggestion
    // Name should contain SHA256 hash in hex (minus last char to stay under BPF_OBJ_NAME_LEN).
```

File: tools/bpf2c/bpf2c.cpp [hunk 2]
```diff
@@ -1,10 +1,10 @@
 // Copyright (c) eBPF for Windows contributors
 // SPDX-License-Identifier: MIT
 
+#include "../../libs/shared/hash.h"
```
Comment (dthaler) [inline tools/bpf2c/bpf2c.cpp:4]: Rather than putting the path here, it's better to put it in the include path in the vcxproj file.

Comment (mikeagun) [reply tools/bpf2c/bpf2c.cpp]: updated

**Review by dthaler — APPROVED** [review]

**Pull request merged**